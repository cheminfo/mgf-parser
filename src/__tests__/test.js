import { readFileSync } from 'fs';
import { join } from 'path';

import max from 'ml-array-max';
import sum from 'ml-array-sum';

import { parse } from '..';

let data = readFileSync(join(__dirname, '../../data/test.mgf'), 'utf8');
let corruptedData = readFileSync(
  join(__dirname, '../../data/lackingEqual.mgf'),
  'utf8',
);
let otherEntryTypeData = readFileSync(
  join(__dirname, '../../data/wrongEntryType.mgf'),
  'utf8',
);

let twoEntryTypesData = readFileSync(
  join(__dirname, '../../data/twoEntryTypes.mgf'),
  'utf8',
);

let basicSpectrumData = readFileSync(
  join(__dirname, '../../data/basicSpectrum.mgf'),
  'utf8',
);

let tabDelimitedData = readFileSync(
  join(__dirname, '../../data/tabDelimitedSpectrum.mgf'),
  'utf8',
);

describe('index', () => {
  it('parse test database', () => {
    let result = parse(data);
    expect(result).toHaveLength(1);
    expect(result).toMatchSnapshot();
    // console.log(result);
  });
  it('parse tab delimited spectrum', () => {
    let result = parse(tabDelimitedData);
    expect(parse(data)).toHaveLength(1);
    expect(result[0].data.x).toHaveLength(214);
    // console.log(result);
  });
  it('metadata corrupted (lacking "=")', () => {
    expect(() => parse(corruptedData)).toThrow(
      'Parsing error at line number 7',
    );
  });
  it('other entry type', () => {
    expect(parse(otherEntryTypeData)[0].kind).toBe('PEPTIDES');
  });
  // testing options
  it('default options (sortX = false and uniqueX = false)', () => {
    let result = parse(basicSpectrumData);
    expect(result).toStrictEqual([
      {
        data: { x: [4, 5, 5, 1, 2, 4], y: [44, 55, 5, 111, 32, 64] },
        kind: 'IONS',
        meta: {},
      },
    ]);
  });
  it('filter entry type IONS', () => {
    let result = parse(twoEntryTypesData, { recordTypes: 'IONS' });
    expect(result).toHaveLength(1);
    expect(result[0].kind).toBe('IONS');
  });
  it('sortX = true and uniqueX = false', () => {
    let result = parse(basicSpectrumData, { sortX: true, uniqueX: false });
    expect(result).toStrictEqual([
      {
        data: { x: [1, 2, 4, 4, 5, 5], y: [111, 32, 44, 64, 55, 5] },
        kind: 'IONS',
        meta: {},
      },
    ]);
  });
  it('sortX = true and uniqueX = true', () => {
    let result = parse(basicSpectrumData, { sortX: true, uniqueX: true });
    expect(result).toStrictEqual([
      {
        data: { x: [1, 2, 4, 5], y: [111, 32, 108, 60] },
        kind: 'IONS',
        meta: {},
      },
    ]);
  });
  it('sortX = false and uniqueX = true', () => {
    let result = parse(basicSpectrumData, { uniqueX: true });
    expect(result).toStrictEqual([
      {
        data: { x: [1, 2, 4, 5], y: [111, 32, 108, 60] },
        kind: 'IONS',
        meta: {},
      },
    ]);
  });
  it('normedY = true', () => {
    let result = parse(basicSpectrumData, { normedY: true });
    expect(sum(result[0].data.y)).toBe(1);
  });
  it('maxY = 100', () => {
    let result = parse(basicSpectrumData, { maxY: 100 });
    expect(max(result[0].data.y)).toBe(100);
  });
  it('should throw error (normedY = true, maxY = 100', () => {
    expect(() =>
      parse(basicSpectrumData, { maxY: 100, normedY: true }),
    ).toThrow('Option maxY must be undefined if normedY is true');
  });
});
