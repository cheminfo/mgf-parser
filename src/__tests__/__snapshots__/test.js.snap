// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`index parse test database 1`] = `
Array [
  Object {
    "data": Object {
      "x": Array [
        41.03912516,
        51.0234751,
        53.03912516,
        57.07042529,
        71.04968984,
        77.03912516,
        79.05477522,
        81.03403978,
        85.06533991,
        93.07042529,
        95.04968984,
        95.08607535,
        97.10172542,
        107.0496898,
        109.0653399,
        131.0496898,
        133.0653399,
        135.08099,
        141.0915547,
        143.1072047,
        147.08099,
        163.0759046,
        165.0915547,
        179.1072047,
        187.0759046,
        199.1334195,
        215.1072047,
        259.1334195,
        261.1490695,
        271.1334195,
        273.1490695,
        275.1283341,
        275.1647196,
        277.1439842,
        277.1803697,
        279.1596342,
        289.1439842,
        289.1803697,
        291.1596342,
        293.1752843,
        295.1909343,
        299.1647196,
        301.1803697,
        305.1752843,
        307.1909343,
        313.1803697,
        315.1596342,
        317.1752843,
        319.1545488,
        319.1909343,
        321.1701989,
        323.185849,
        331.1909343,
        333.1701989,
        349.201499,
      ],
      "y": Array [
        0.8087251542,
        3.32753245,
        2.567047121,
        2.221750186,
        2.248179447,
        1.829006615,
        1.902007135,
        1.488314872,
        4.6347884385,
        2.7620138170999997,
        0.653485705,
        1.408259117,
        1.599446795,
        5.5722129241,
        6.4124522102,
        1.621699342,
        1.700097586,
        2.638009055,
        0.2428717614,
        0.5688016668,
        3.3303456194,
        1.0664100554,
        0.8715727915,
        2.803534357,
        2.7761789844,
        0.2266414054,
        1.848992754,
        1.706102891,
        0.5332527875,
        9.5851991877,
        6.195687958000001,
        2.75850236,
        3.9493072280000003,
        0.4641393593,
        0.5523075808,
        0.5249912699,
        4.690200860999999,
        1.461952458,
        2.3208544498,
        2.1472810351,
        0.190602914,
        8.072470757,
        6.1067024053,
        0.4530712565,
        0.1661152974,
        79.4106497,
        4.8992273722,
        1.6569410315000002,
        0.533269375,
        2.0372253968000003,
        0.5650030571,
        0.5952336891,
        72.475414317,
        0.3957317926,
        26.422184839,
      ],
    },
    "kind": "IONS",
    "meta": Object {
      "CHARGE": "1+",
      "EXACTMASS": "348.193674",
      "FILENAME": "PMA_065444",
      "INCHI": "InChI=1/C20H28O5/c1-9-5-13-12(15(9)22)6-11(8-21)7-14-16-18(3,4)20(16,25)17(23)10(2)19(13,14)24/h5,7,10,14-17,21-25H,6,8H2,1-4H3/t10-,14+,15+,16-,17-,19+,20-/m1/s1",
      "INCHIKEY": "CECQLSSRZIBODZ-MFYSWCGQSA-N",
      "IONMODE": "Positive",
      "MOLECULAR_FORMULA": "C20H28O5",
      "NAME": "UNPD159661",
      "PEPMASS": "349.20095",
      "SCANS": "1",
      "SEQ": "*..*",
      "SMILES": "CC1C(O)C2(O)C(C3C=C(CO)CC4=C(C=C(C)C4O)C13O)C2(C)C",
      "UNPD_ID": "UNPD159661",
    },
  },
]
`;
